### PR TITLE
NMS-13842: Extend SnmpMetadataProvisioningAdapter to support exact OIDs

### PR DESCRIPTION
**Pull request by Naicisum**
NMS-13842: Extend SnmpMetadataProvisioningAdapter to support exact OIDs

### All Contributors

* [x] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### Contribution Checklist

* Please [make an issue in the OpenNMS issue tracker](https://issues.opennms.org) if there isn't one already.<br />Once there is an issue, please:
  1. update the title of this PR to be in the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
  4. once you've created this PR, please link to it in a comment in the JIRA issue
  Don't worry if this sounds like a lot, we can help you get things set up properly.
* If this is a new or updated feature, is there documentation for the new behavior?
* If this is new code, are there unit and/or integration tests?
* If this PR targets a `foundation-*` branch, does it try to avoid changing files in `$OPENNMS_HOME/etc/`?

### What's Next?

A PR should be assigned at least 2 reviewers.  If you know that someone would be a good person to review your code, feel free to add them.

If you need help making additions or changes to the documentation related to your changes, please let us know.

In any case, if anything is unclear or you want help getting your PR ready for merge, please don't hesitate to say something in the comments here,
or in [the #opennms-development chat channel](https://chat.opennms.com/opennms/channels/opennms-development).

Once reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the OGP) you are welcome to merge the PR when you're ready.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13842

### CODE REVIEW

**Review by christianpape — DISMISSED** [review]
Looks good to me.

**Comment by fooker** [comment]
@christianpape why appending `.0` at all? I would expect the thing to query the exact value anyways (if not querying for indexed values)?

File: docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc [hunk 1]
```diff
@@ -28,6 +28,10 @@ The example above will query nodes with System Object Ids matching the regular e
 The data will be searched in the SNMP sub-tree `.1.2.3.4` and be stored in the node's metadata context `snmp`.
 The top-level element will be named `myHardware`.
 
+The provisioning adapter automatically appends the `.0` value to single entry OIDs per normal SNMP specification.
+There are edge cases where we will need to specify the entire OID statically to retrieve a single piece of data.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc:32]: This sentence sounds like a note to OpenNMS. Could we change to:

In some edge cases, you will need to specify the entire OID statically to retrieve a single piece of data. 

Can we specify what those edge cases might be, or how someone can determine when they would need to specify the entire OID, or will an error message appear if they do not?

File: docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc [hunk 2]
```diff
@@ -28,6 +28,10 @@ The example above will query nodes with System Object Ids matching the regular e
 The data will be searched in the SNMP sub-tree `.1.2.3.4` and be stored in the node's metadata context `snmp`.
 The top-level element will be named `myHardware`.
 
+The provisioning adapter automatically appends the `.0` value to single entry OIDs per normal SNMP specification.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc:31]: "single-entry OIDs"

please add hyphen.

**Review by Bonrob2 — COMMENTED** [review]
Minor doc changes and question.

File: docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc (hunk 1 above)
Comment (christianpape) [reply docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc]: A SNMP scalar value has a .0 at the end of its OId, e.g. ".1.2.3.4.5.0". A list of values has its instance identifier at the end, e.g. ".1.2.3.4.5.1.2.3" where "1.2.3" is the instance id.

File: docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc (hunk 2 above)
Comment (christianpape) [reply docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc]: done

**Review by Bonrob2 — APPROVED** [review]
LGTM! Thanks.

File: docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc (hunk 1 above)
Comment (Bonrob2) [reply docs/modules/reference/pages/provisioning/adapters/snmp-metadata.adoc]: Thanks for clarifying.

**Review by christianpape — APPROVED** [review]
All good.

**Comment by christianpape** [comment]
> @christianpape why appending `.0` at all? I would expect the thing to query the exact value anyways (if not querying for indexed values)?

I tried to align the config as most as possible to MIB definitions.  So, scalar values indices (.0) are implicit added.

**Review by RangerRick — APPROVED** [review]
simple fix, looks good to me

**Pull request merged**